fluent-syntax/src/ast.js

```diff
@@ -5,8 +5,86 @@
  * Annotation.
  *
  */
-class BaseNode {
+export class BaseNode {
   constructor() {}
+
+  equals(other, ignoredFields = ["span"]) {
+    const thisKeys = new Set(Object.keys(this));
+    const otherKeys = new Set(Object.keys(other));
+    if (ignoredFields) {
+      for (const fieldName of ignoredFields) {
+        thisKeys.delete(fieldName);
+        otherKeys.delete(fieldName);
+      }
+    }
+    if (thisKeys.size !== otherKeys.size) {
+      return false;
+    }
+    for (const fieldName of thisKeys) {
+      if (!otherKeys.has(fieldName)) {
+        return false;
+      }
+      const thisVal = this[fieldName];
+      const otherVal = other[fieldName];
+      if (typeof thisVal !== typeof otherVal) {
+        return false;
+      }
+      if (thisVal instanceof Array) {
+        if (thisVal.length !== otherVal.length) {
+          return false;
+        }
+        // Sort elements of order-agnostic fields to ensure the
+        // comparison is order-agnostic as well. Annotations should be
+        // here too but they don't have sorting keys.
+        if (["attributes", "variants"].includes(fieldName)) {
+          thisVal.sort(sortingKeyCompare);
+          otherVal.sort(sortingKeyCompare);
```

Comment:
Should we make copies of both arrays here, just in case someone tries to serialize the AST (or work with it in some other fashion) after running `equals` on it?

Reply:
I'd rather avoid the cost for that.

Reply:
Do I understand it correctly, that not doing it introduces a bug: the AST is mutated when `equals` is called?

Reply:
Thinking about this, we should just not make nodes equal if the order changes.

I filed https://github.com/projectfluent/python-fluent/issues/111 on this.